### ISSUE TITLE
Proposal: Allow ARG before FROM (Dockerfile templates)

### ISSUE THREAD

**Issue by alias1** [docker/docker#32457]
Currently it doesn't seem possible to use the `ARG` statement before `FROM` in a `Dockerfile`:

```
Sending build context to Docker daemon 2.598 MB
Step 1/3 : ARG foo
Please provide a source image with `from` prior to commit
```

I've noticed a pattern with a number of Dockerfiles where they are using a template `Dockerfile`, and then parsing it with some shell scripts to generate the specific Dockerfiles for each image: eg. The file for ubuntu, alpine, etc.

I was thinking it might be interesting to solve this problem by using the [build args](https://docs.docker.com/engine/reference/builder/#arg) to make a sort of 'Dockerfile template'. 

Below is a contrived example:

```
# Dockerfile
ARG myfrom
ARG myplatform
FROM ${myfrom}

COPY scripts/ /scripts/

RUN /scripts/foo.${myplatform}.sh

CMD ["/bin/sh"]
```

I could then pass through the build args like:

```
# Alpine
docker build --build-arg myfrom="usualoma/ruby-with-therubyracer:2.3.0-alpine" --build-arg myplatform=alpine .

# Ubuntu
docker build --build-arg myfrom="ruby:2.3" --build-arg myplatform=debian .
```

There seem to be other options currently available, but require 3rd party tools, eg:
* https://www.npmjs.com/package/dockerfile-template

It would also be pretty useful if we were able to `RUN` a build script without having to copy it into the container first. Eg. `RUN-FROMCONTEXT foo.sh`.

**Comment by AkihiroSuda** [comment]
Work in progress: https://github.com/docker/docker/pull/31352

let me close this issue

**Comment by alias1** [comment]
Oh, must have missed that. Thanks!

**Comment by roryodonnell** [comment]
```
rory@ubuntu:~/git/broker-hdsm$ make build
Building PIP package...
/usr/src/broker-hdsm
Building Docker image...
Sending build context to Docker daemon 8.777 MB
Step 1/12 : ARG BASE_IMAGE
Please provide a source image with `from` prior to commit
Makefile:84: recipe for target 'build' failed
make: *** [build] Error 1

```

**Comment by thaJeztah** [comment]
@roryodonnell that's expected; Docker 1.13 is really old, and has reached end of life in March of last year. This feature is available in Docker 17.05 and up so any _current_ version of Docker should have this feature.

**Comment by adamkdean** [comment]
FWIW this was introduced in [17050-ce-2017-05-04](https://docs.docker.com/release-notes/docker-ce/#17050-ce-2017-05-04).
